crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+        trace!("Couldn't find the event attached to the latest receipt; looking if the past latest known receipt refers to a new event...");
+        if find_and_count_events(&receipt_event_id, user_id, new_events.iter(), room_info) {
+            // We found the event to which the previous receipt attached to, so our work is
+            // done here.
+            return Ok(());
+        }
+    }
+
+    // If we haven't returned at this point, it means that either we had no previous
+    // read receipt, or the previous read receipt was not attached to any new
+    // event.
+    //
+    // In that case, accumulate all events as part of the current batch, and wait
+    // for the next receipt.
+    trace!("All other ways failed, including all new events for the receipts count.");
+    for event in new_events {
+        if event.push_actions.iter().any(ruma::push::Action::is_highlight) {
+            room_info.notification_counts.highlight_count += 1;
+        }
+        if marks_as_unread(&event.event, user_id) {
+            room_info.notification_counts.notification_count += 1;
+        }
+    }
+
+    Ok(())
+}
+
+/// Try to find the event to which the receipt attaches to, and if found, will
+/// update the notification count in the room.
+///
+/// Returns a boolean indicating if it's found the event and updated the count.
+fn find_and_count_events<'a>(
+    receipt_event_id: &EventId,
+    user_id: &UserId,
+    events: impl Iterator<Item = &'a SyncTimelineEvent>,
+    room_info: &mut RoomInfo,
+) -> bool {
+    let mut counting_receipts = false;
+    for event in events {
+        if counting_receipts {
+            for action in &event.push_actions {
+                if action.is_highlight() {
+                    room_info.notification_counts.highlight_count += 1;
+                }
+                if action.should_notify() && marks_as_unread(&event.event, user_id) {
+                    room_info.notification_counts.notification_count += 1;
+                }
+            }
+        } else if let Ok(Some(event_id)) = event.event.get_field::<OwnedEventId>("event_id") {
+            if event_id == receipt_event_id {
+                // Bingo! Switch over to the counting state, after resetting the
+                // previous counts.
+                trace!("Found the event the receipt was referring to! Starting to count.");
+                room_info.notification_counts = Default::default();
+                counting_receipts = true;
+            }
+        }
+    }
+    counting_receipts
+}
+
+/// Is the event worth marking a room as unread?
+fn marks_as_unread(event: &Raw<AnySyncTimelineEvent>, user_id: &UserId) -> bool {
+    let event = match event.deserialize() {
+        Ok(event) => event,
+        Err(err) => {
+            tracing::debug!(
+                "couldn't deserialize event {:?}: {err}",
+                event.get_field::<String>("event_id").ok().flatten()
+            );
+            return false;
+        }
+    };
```

Comment:
This pattern is a good candidate for `let Ok(event) = event.deserialize() else { debug!(…); return false; }` if you like it.

Reply:
That would not allow us to display the error, so I'll keep it like that :+1: